query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+            _ => bail!(ErrorKind::UnknownFunction(where_fn.operator.clone())),
+        }
+    }
+
+    #[allow(unused_variables)]
+    pub fn apply_ground<'s>(&mut self, schema: &'s Schema, where_fn: WhereFn) -> Result<()> {
+        if where_fn.args.len() != 2 {
+            bail!(ErrorKind::InvalidNumberOfArguments(where_fn.operator.clone(), where_fn.args.len(), 2));
+        }
+
+        let mut args = where_fn.args.into_iter();
+
+        // TODO: process source variables.
+        match args.next().unwrap() {
+            FnArg::SrcVar(SrcVar::DefaultSrc) => {},
+            _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "source variable".into(), 0)),
+        }
+
+        if where_fn.binding.is_empty() {
+            // The binding must introduce at least one bound variable.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::NoBoundVariable));
+        }
+
+        if !where_fn.binding.is_valid() {
+            // The binding must not duplicate bound variables.
+            bail!(ErrorKind::InvalidBinding(where_fn.operator.clone(), BindingError::RepeatedBoundVariable));
+        }
+
+        // Turn input constant into a matrix, ready to feed into a VALUES clause.
+        let constant = args.next().unwrap();
+        let rows = { // Scope borrow of self.
+            let operator = where_fn.operator;
+            let lookup = |in_var| {
+                if !self.input_variables.contains(&in_var) {
+                    bail!(ErrorKind::UnboundVariable((*in_var.0).clone()));
+                }
+                match self.bound_value(&in_var) {
+                    Some(ref in_value) => Ok(in_value.clone()),
+                    None => bail!(ErrorKind::UnboundVariable((*in_var.0).clone())),
+                }
+            };
+
+            match where_fn.binding {
+                Binding::BindScalar(_) => constant.into_unit_matrix(&lookup)?,
+                Binding::BindTuple(_) => constant.into_row_matrix(&lookup)?,
+                Binding::BindColl(_) => constant.into_column_matrix(&lookup)?,
+                Binding::BindRel(_) => constant.into_block_matrix(&lookup)?,
+            }
+        };
+
+        // Typecheck and drop columns that are not actually bound.  By restricting to homogeneous
+        // columns, we greatly simplify projection.  In the future, we could loosen this
+        // restriction, at the cost of projecting (some) value type tags.  If and when we want to
+        // algebrize in two phases and allow for late-binding input variables, we'll probably be
+        // able to loosen this restriction with little penalty.
+        let variables: Vec<Option<Variable>> = where_fn.binding.variables();
+        let (values, column_types) = reduced_rows_and_column_types(rows, &variables.iter().map(|x| x.is_some()).collect())?;
+        let names: Vec<Variable> = variables.into_iter().filter_map(|x| x).collect();
+
+        let named_values = ComputedTable::NamedValues {
```

Comment:
I wonder if this should be called `NamedRel` or `NamedStaticRel` instead — what you're defining here is a relation to join into the query. That just happens to be implemented on the SQL side as `VALUES`, but it could also be implemented as `IN` (for coll), as simple substitution (for scalar), or even as a materialized table.

The `Static` bit might indicate that there's a different kind of rel, which is computed from a function…